modes/parkour/leaderboard.lua

```diff
@@ -123,36 +125,65 @@ onEvent("GameDataLoaded", function(data)
 		end
 
 		local new_reset = os.date("%d/%m/%Y", ts - now.wday * 24 * 60 * 60 * 1000)
-		if new_reset ~= timed_maps.week.last_reset then
+		if new_reset ~= data.weekly.ts then
 			if new_reset == "28/02/2021" then
 				translatedChatMessage("double_maps_start")
 			elseif new_reset == "07/03/2021" then
 				translatedChatMessage("double_maps_end")
 			end
 
-			if #data.weekranking > 2 and data.weekranking[1][3] > 30 then
-				sendPacket(
-					"common", 4,
-					timed_maps.week.last_reset .. "\000" .. os.date("%d/%m/%Y", ts - 24 * 60 * 60 * 1000) ..
-					"\000" .. data.weekranking[1][4] .. "\000" .. data.weekranking[1][2] .. "\000" .. data.weekranking[1][3] ..
-					"\000" .. data.weekranking[2][4] .. "\000" .. data.weekranking[2][2] .. "\000" .. data.weekranking[2][3] ..
-					"\000" .. data.weekranking[3][4] .. "\000" .. data.weekranking[3][2] .. "\000" .. data.weekranking[3][3]
-				)
-				channelHandler(true) -- force send
+			if #weekleaderboard > 2 and weekleaderboard[1][3] > 30 then
+				data.weekly.lw = data.weekly.cw
+				data.weekly.cw = {weekleaderboard[1][1], weekleaderboard[2][1], weekleaderboard[3][1]}
 			end
 
+			data.weekly.ts = new_reset
+			data.weekly.ranks = {}
+		end
+
+		if timed_maps.week.last_reset ~= new_reset then
 			timed_maps.week.last_reset = new_reset
-			timed_maps.week.next_reset = os.date("%d/%m/%Y", ts + (7 - now.wday) * 24 * 60 * 60 * 1000)
 
 			for player, data in next, players_file do
-				data.week = {0, new_reset}
+				if data.week[2] ~= new_reset then
+					data.week = {0, new_reset}
+				end
 			end
 
-			data.weekranking = {}
+			tfm.exec.chatMessage("<j>The weekly leaderboard has been reset.")
 		end
 
-		weekleaderboard = data.weekranking
+		weeklyfile = data.weekly
+		weekleaderboard = data.weekly.ranks
 
 		checkPlayersPosition(true)
 	end
 end)
+
+local function in_table(value, tbl)
+	for _, v in ipairs(tbl) do
+		if v == value then
+			return true
+		end
+	end
+	return false
+end
+
+local function checkWeeklyWinners(player, data)
+	if not weeklyfile.lw or not weeklyfile.cw then return end
+
+	local roomPlayer = room.playerList[player]
+	local lwIndex = in_table(tonumber(roomPlayer.id), weeklyfile.lw)
+	local cwIndex = in_table(tonumber(roomPlayer.id), weeklyfile.cw)
+
+	if not lwIndex and not cwIndex then return end
```

Comment:
This seems a little resource intensive; changing it to a dictionary might help. I also thought it'd help to cleanup the file a little; but I'm not exactly sure whether or not you're doing it in another place.
```suggestion
	local id = tostring(room.playerList[player].id)
	if not weeklyfile.lw[id] and not weeklyfile.cw[id] then return end
	weeklyfile.lw[id] = nil
	weeklyfile.cw[id] = nil
```

```lua
{
	name = "cw",
	type = "dictionary",
	objects = {
		type = "boolean" 
	},
},
{
	name = "lw",
	type = "dictionary",
	objects = {
		type = "boolean" 
	},
},
```